Cover case when target branch is pointing to release branch

jhadvig opened the pull request:
Fixing use case when either openshift-ansible or origin target branch was pointing to a release branch and we ended up installing/upgrading different versions of those two pkgs.
[Example](https://ci.openshift.redhat.com/jenkins/job/test_pull_request_openshift_ansible_extended_conformance_install_update/270/consoleFull): 
```
+ source ORIGIN_VARS
++ ORIGIN_INSTALL_VERSION=1.5.1-1.el7
++ ORIGIN_INSTALL_MINOR_VERSION=5
++ ORIGIN_UPGRADE_RELEASE_VERSION=3.6.0-0.alpha.1.1013.322171b
++ ORIGIN_UPGRADE_RELEASE_MINOR_VERSION=6
+ source OPENSHIFT_ANSIBLE_VARS
++ ATOMIC_OPENSHIFT_UTILS_INSTALL_VERSION=3.4.94-1.git.0.fe1fe27.el7
++ ATOMIC_OPENSHIFT_UTILS_INSTALL_MINOR_VERSION=4
++ ATOMIC_OPENSHIFT_UTILS_UPGRADE_RELEASE_VERSION=3.5.79-1.git.0.f5f10f5.el7
++ ATOMIC_OPENSHIFT_UTILS_UPGRADE_RELEASE_MINOR_VERSION=5
```

For that reason I've added an optional `--dependency_branch` flag which represents the target branch of the dependency repository. So for example if the PR is targeting  `OPENSHIFT_ANSIBLE_TARGET_BRANCH=release-1.5` we have to install `origin-1.4-x` and upgrade to `origin-1.5.x` and vice versa (if ORIGIN_TARGET_BRANCH=release-1.5 the install version of atomic-openshift-utils should be `1.4.x` and the upgrade version `1.5.x`) so the versions are matching.

Updated  `determine_install_upgrade_version.py` script && its TCs && install_udatte job

@stevekuznetsov PTAL

stevekuznetsov commented on sjb/hack/determine_install_upgrade_version.py:
```diff
@@ -64,14 +63,29 @@ def determine_search_version(pkg_name, pkg_version):
+	#          ORIGIN_UPGRADE_RELEASE_VERSION=1.5.1-1.el7
+	#          ORIGIN_UPGRADE_RELEASE_MINOR_VERSION=5
+	#
+	if args.dependency_branch.startswith("release"):
```
```py
if args.dependency_branch.startswith("release") or args.dependency_branch.startswith("enterprise"):
```

jhadvig commented:
@stevekuznetsov comment addressed + updated the logic a bit

jhadvig commented on sjb/hack/determine_install_upgrade_version.py:
```diff
@@ -84,9 +98,21 @@ def determine_search_version(pkg_name, pkg_version):
+
+	# If dependency target branch is specified for a release or an enterprise release, determine the upgrade version
+	# by listing all the versions which major and minor version match with the release version and from the list pick
+	# the last one. Otherwise upgrade to the version of the input_pkg
```
This is what I've added. If the dependency is set to a release version determine the upgrade version otherwise use the input_pkg version

stevekuznetsov replied:
Confused -- we always upgrade to the version being built

jhadvig replied:
Even if:
OPENSHIFT_ANSIBLE_TARGET_BRANCH = release-1.5
ORIGIN_TARGET_BRANCH = master
in case of
https://ci.openshift.redhat.com/jenkins/job/test_pull_request_openshift_ansible_extended_conformance_install_update/280/parameters/

in this case we can could even skip the building of origin.. right ?

stevekuznetsov replied:
Oh, I see. So this is only for the _other_ package. Yes, I agree then

jhadvig replied:
yeah.. sorry for the confusion, took me a bit to realize whats the issue as well :smile:

jhadvig replied:
will update the PR and skipp the building of origin in case the `OPENSHIFT_ANSIBLE_TARGET_BRANCH=release-1.*`  or skip the building of openshift-ansible pkgs if `ORIGIN_TARGET_BRANCH=release-1.*` is set

jhadvig commented:
@stevekuznetsov I've updated the PR and now we are skipping the origin build if the `OPENSHIFT_ANSIBLE_TARGET_BRANCH` is not set to master. Not skipping the build of openshift-ansible pkgs since we need the `OPENSHIFT_ANSIBLE_BUILT_VERSION`, which is set during that process and would rather keep it then mock it. The build of those pkgs doesnt take that long in compare with the build of origin.
PTAL

stevekuznetsov commented:
That seems reasonable

jhadvig commented:
@stevekuznetsov please let me know when you merge this. Will push the changes into jenkins. thnaks !

kargakis commented on sjb/config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -37,7 +37,7 @@ extensions:
         echo $pkg_name > PKG_NAME
         echo "openshift-ansible openshift-ansible-callback-plugins openshift-ansible-docs openshift-ansible-filter-plugins openshift-ansible-lookup-plugins openshift-ansible-playbooks openshift-ansible-roles" > OPENSHIFT_ANSIBLE_PKGS
-        sudo python sjb/hack/determine_install_upgrade_version.py $( cat OPENSHIFT_ANSIBLE_BUILT_VERSION ) > OPENSHIFT_ANSIBLE_VARS
+        sudo python sjb/hack/determine_install_upgrade_version.py $( cat OPENSHIFT_ANSIBLE_BUILT_VERSION ) --dependency_branch ${ORIGIN_TARGET_BRANCH} > OPENSHIFT_ANSIBLE_VARS
```
How come isn't this affected by https://github.com/openshift/aos-cd-jobs/issues/315?

jhadvig replied:
Added the variable with `type: "forward_parameters"` action

jhadvig commented:
@stevekuznetsov @kargakis I've updated the PR. I've added the `ORIGIN_TARGET_BRANCH` and `OPENSHIFT_ANSIBLE_TARGET_BRANCH` variables with `type: "forward_parameters"` action. And tested the if they are present in running scripts and it was working.
PTAL

kargakis commented on sjb/config/test_cases/test_branch_origin_extended_conformance_install_update.yml:
```diff
@@ -5,6 +5,10 @@ extensions:
   actions:
+    - type: "forward_parameters"
+      parameters:
+        - ORIGIN_TARGET_BRANCH
```
Hrm, this doesn't solve https://github.com/openshift/aos-cd-jobs/issues/315 holistically for all jobs or even all repos for this job :(

jhadvig replied:
it wont, but at least until we fix this globally the PRs that point to release branches wont fail.

kargakis replied:
Fine by me

stevekuznetsov commented:
Do we have a test job for this?

jhadvig commented:
@stevekuznetsov I've update the PR and added additional commit that will deal with the wrongly tagged `origin-3.6.0-0.0.alpha.0.1 package`  + added aadditional TCs for this case. The test for these commits is running here https://ci.openshift.redhat.com/jenkins/view/All/job/jhadvig-test/6/console

stevekuznetsov commented:
@jhadvig ok let me know when you are happy with what that job did

jhadvig commented:
@stevekuznetsov the test is green. Did some additional manual tests for different use-cases and everything should be as should now. Lets merge this !

stevekuznetsov commented:
You push it out, ok?
